solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java

```diff
@@ -191,19 +213,79 @@ public void uploadConfig(Path dir, String configName) throws IOException {
   }
 
   @Override
-  public void uploadFileToConfig(String fileName, String configName) throws IOException {
-    Path path = Paths.get(fileName);
-    if (!Files.exists(path)) {
-      throw new IOException("File path " + path + " does not exist");
+  public void createFilePathInConfig(String configName, String fileName, boolean failOnExists) throws IOException {
+    try {
+      zkClient.makePath(CONFIGS_ZKNODE + "/" + configName + "/" + fileName, failOnExists,true);
+    } catch (KeeperException | InterruptedException e) {
+      throw new IOException("Error creating file path in config", SolrZkClient.checkInterrupted(e));
+    }
+  }
+
+  @Override
+  public void uploadFileToConfig(String configName, String fileName, byte[] data) throws IOException {
+    try {
+      zkClient.makePath(CONFIGS_ZKNODE + "/" + configName + "/" + fileName, data, true);
+    } catch (KeeperException | InterruptedException e) {
+      throw new IOException("Error creating file in config", SolrZkClient.checkInterrupted(e));
+    }
+  }
+
+  @Override
+  public void uploadFileToConfig(String configName, String fileName, byte[] data, boolean failOnExists) throws IOException {
+    String filePath = CONFIGS_ZKNODE + "/" + configName + "/" + fileName;
+    try {
+      zkClient.makePath(filePath, data, CreateMode.PERSISTENT, null, failOnExists, true);
+    } catch (KeeperException.NodeExistsException nodeExistsException) {
+      throw new SolrException(
+          SolrException.ErrorCode.BAD_REQUEST,
+          "The path " + filePath + " for configSet " + configName + " already exists. In order to overwrite, provide overwrite=true or use an HTTP PUT with the V2 API.");
+    } catch (KeeperException | InterruptedException e) {
+      throw new IOException("Error creating file in config", SolrZkClient.checkInterrupted(e));
+    }
+  }
+
+  @Override
+  public void setConfigMetadata(String configName, Map<Object, Object> data) throws IOException {
+    try {
+      zkClient.makePath(CONFIGS_ZKNODE + "/" + configName, Utils.toJSON(data), true);
+    } catch (KeeperException | InterruptedException e) {
+      throw new IOException("Error setting config metadata", SolrZkClient.checkInterrupted(e));
+    }
+  }
+
+  @Override
+  public void setConfigMetadata(String configName, byte[] data) throws IOException {
```

Comment:
No; I think the API of ConfigSetService should keep this higher level than bytes.  The Map one is perfect.